[jit] guard against inheritance on torchscript classes

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28409 [jit] submodule swapping via module interface
* #28408 [jit] introduce module interface declaration
* **#28407 [jit] guard against inheritance on torchscript classes**

Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284310](https://our.internmc.facebook.com/intern/diff/D18284310)

wanchaol committed:
[jit] guard against inheritance on torchscript classes

Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
rebase with master on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on torch/jit/__init__.py:
```diff
@@ -1250,6 +1250,9 @@ def forward(self, input):
             raise RuntimeError("TorchScript classes must be new-style classes. "
                                "Please inherit from 'object'")
+        if len(obj.mro()) > 2:
+            raise RuntimeError("TorchScript classes does not support inheritance or any polymorphism strategy yet. "
```
"or any polymorphism strategy" is not correct. Interfaces are a form of polymorphism.  The first part of the message is fine.

zdevito commented on torch/jit/__init__.py:
```diff
@@ -1250,6 +1250,9 @@ def forward(self, input):
                                "Please inherit from 'object'")
+        if len(obj.mro()) > 2:
+            raise RuntimeError("TorchScript classes does not support inheritance or any polymorphism strategy yet. "
+                               "Please directly inherit from 'object'. File a issue if you want this. ")
```
Let's keep 'File an issue' for bugs only. It implies implementing this would be easy when it is in face very difficult, which is why we haven't implemented it.

zdevito reviewed (approved):
Looks good, small phrasing comment.

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] guard against inheritance on torchscript classes"


Summary:

Given that we do not have support for inheitance or any polymorphism
strategy yet, we should guard against user from using it until we get
the full support so that user won't confuse by the weird behaviors.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284310](https://our.internmc.facebook.com/intern/diff/D18284310)

[ghstack-poisoned]
